### ISSUE TITLE
Add OCRmyPDF / Teseract as alternative for pdftotext / Azure Machine Vision

### ISSUE THREAD

**Issue by stv0g** [meine-stadt-transparent/meine-stadt-transparent#751]
See: https://ocrmypdf.readthedocs.io/en

I have made pretty good experiences while using it with Paperless-ng:

https://github.com/jonaswinkler/paperless-ng/blob/master/src/paperless_tesseract/parsers.py

OCRmyPDF also provides a simple Docker-based webservice which we could simple add to the docker-compose setup.

In Kubernetes deployments this service could be scaled up to multiple instances to distribute the load..

**Comment by konstin** [comment]
I did some research into ocr tools a while age, the main problem was that none of them was installable through pip and I didn't want to add a queuing system just for that. But in general I'm happy to add more backends for that

**Comment by stv0g** [comment]
I have now a first version based on [OCRmyPDF](https://ocrmypdf.readthedocs.io/) / [pdfminer.six](https://pdfminersix.readthedocs.io/) working which just needs some more cleanup before I can push it 🥳 

However, I would like to parallelize the huge task of OCRing >10k files.
Currently my best bet is: [django-q](https://django-q.readthedocs.io/en/latest/) which could also be quite useful for the other periodic tasks. It also fits quite nicely in my Kubernetes architecture as I can distribute the workload across multiple nodes..

**Comment by stv0g** [comment]
This also means I have abondoned the idea of using the OCRmyPDF web-service and rather use django-q for this.

**Comment by konstin** [comment]
> However, I would like to parallelize the huge task of OCRing >10k files.

FWIW the import_files job can already run in parallel, we already need that for text extraction and person/geotagging

> This also means I have abondoned the idea of using the OCRmyPDF web-service and rather use django-q for this.

django-q looks nice! I think defaulting to multiprocessing while supporting redis als broker would be nice. The sentry integration is another plus

**Comment by stv0g** [comment]
Now that django-q has been merged into MST (#762), I can proceed to finish the OCR integration.

I have added two more implementations of [`concurrent.Executor`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor):

-  [`SingleThreadExecutor`](https://github.com/odlabac/aachen-transparent/blob/13ceb3558bf2deaacfeb0f8e84fdc7790260509b/importer/executor.py#L59) (running all tasks sequentially in a single thread)
- [`WorkerPoolExecutor`](https://github.com/odlabac/aachen-transparent/blob/13ceb3558bf2deaacfeb0f8e84fdc7790260509b/importer/executor.py#L74) (using Django-Q)

in addition to the already existing executors provided by the [`concurrent`](https://docs.python.org/3/library/concurrent.futures.html#executor-objects) package:

- `ThreadPoolExecutor`
- `ProcessPoolExecutor`

This gives a nice generic way of executing various tasks.
E.g. you can select the executor when running import tasks from the CLI via `./manage.py`.

I also added a simple [`run()`](https://github.com/odlabac/aachen-transparent/blob/13ceb3558bf2deaacfeb0f8e84fdc7790260509b/importer/executor.py#L21) function which executes a callback function for each element of an iterable using a given executor.

An added bonus is the integration of this run() function with `tdqm` for a visual feedback to the user in form of a progress bar.

I still need to add tests for the new executors as integrate it in all CLI commands.

**Comment by stv0g** [comment]
On another note:

I am considering storing both the OCR'd and the original files since OCRmyPDF fill mutate the PDFs to add overlay text layers of the OCR'd text.

I have yet to figure our what the best approach would be:
- Extending the File model and added another reference to the OCRd PDF in the S3 store?
   Seems difficult as the S3 store is using the file ID as primary key.
- Creating a second File instance and adding a new field to the Paper / AgendItem models?
- Or just overwriting the original file?

@konstin Do you have any opinion on this?

**Comment by konstin** [comment]
Great to hear that the OCR implementation is progressing! I have a big caveat thought: The original pdfs are official documents and I'd highly recommend not to change them at all. The document originally uploaded is the canonical version that everybody refers to and we should not silently create a second version, there's the danger of unwittingly changing document contents/semantics which must not happen. What we can do (and afaik there's already a switch for that) is to display the text as "alt text" behind the pdf. If you want to also provide a second pdf version feel free to create additional buckets as you like (I treat buckets just like database table), just please make sure the canonical version is still displayed as main version. A second bucket should be the easiest anyway since you can again use the primary key as id so no changes to the file model needed

On a side note, it would be nice MST could otherwise work if the `apt-get install`ed dependencies of the ocr pipeline are missing (not a blocker though
